src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java

```diff
@@ -0,0 +1,861 @@
+     * @param readsDictionary sequence dictionary for the reads
+     * @return a VCF or GVCF writer as appropriate, ready to use
+     */
+    public VariantContextWriter makeVCFWriter( final String outputVCF, final SAMSequenceDictionary readsDictionary ) {
+        VariantContextWriter writer = GATKVariantContextUtils.createVCFWriter(new File(outputVCF), readsDictionary, false);
+
+        if ( hcArgs.emitReferenceConfidence == ReferenceConfidenceMode.GVCF ) {
+            try {
+                writer = new GVCFWriter(writer, hcArgs.GVCFGQBands, hcArgs.genotypeArgs.samplePloidy);
+            } catch ( IllegalArgumentException e ) {
+                throw new UserException.BadArgumentValue("GQBands", "are malformed: " + e.getMessage());
+            }
+        }
+
+        return writer;
+    }
+
+    /**
+     * Writes an appropriate VCF header, given our arguments, to the provided writer
+     *
+     * @param vcfWriter writer to which the header should be written
+     */
+    public void writeHeader( final VariantContextWriter vcfWriter ) {
+        final Set<VCFHeaderLine> headerInfo = new HashSet<>();
+
+        headerInfo.addAll(genotypingEngine.getAppropriateVCFInfoHeaders());
+        // all annotation fields from VariantAnnotatorEngine
+        headerInfo.addAll(annotationEngine.getVCFAnnotationDescriptions());
+        // all callers need to add these standard annotation header lines
+        headerInfo.add(GATKVCFHeaderLines.getInfoLine(GATKVCFConstants.DOWNSAMPLED_KEY));
+        headerInfo.add(GATKVCFHeaderLines.getInfoLine(GATKVCFConstants.MLE_ALLELE_COUNT_KEY));
+        headerInfo.add(GATKVCFHeaderLines.getInfoLine(GATKVCFConstants.MLE_ALLELE_FREQUENCY_KEY));
+        // all callers need to add these standard FORMAT field header lines
+        VCFStandardHeaderLines.addStandardFormatLines(headerInfo, true,
+                VCFConstants.GENOTYPE_KEY,
+                VCFConstants.GENOTYPE_QUALITY_KEY,
+                VCFConstants.DEPTH_KEY,
+                VCFConstants.GENOTYPE_PL_KEY);
+
+        if ( ! hcArgs.doNotRunPhysicalPhasing ) {
+            headerInfo.add(GATKVCFHeaderLines.getFormatLine(GATKVCFConstants.HAPLOTYPE_CALLER_PHASING_ID_KEY));
+            headerInfo.add(GATKVCFHeaderLines.getFormatLine(GATKVCFConstants.HAPLOTYPE_CALLER_PHASING_GT_KEY));
+        }
+
+        // FILTER fields are added unconditionally as it's not always 100% certain the circumstances
+        // where the filters are used.  For example, in emitting all sites the lowQual field is used
+        headerInfo.add(GATKVCFHeaderLines.getFilterLine(GATKVCFConstants.LOW_QUAL_FILTER_NAME));
+
+        if ( emitReferenceConfidence() ) {
+            headerInfo.addAll(referenceConfidenceModel.getVCFHeaderLines());
+        }
+
+        vcfWriter.writeHeader(new VCFHeader(headerInfo, sampleSet));
+    }
+
+    public ActivityProfileState isActive( final FeatureContext features, final ReferenceContext ref, final AlignmentContext context ) {
```

Comment:
I wonder if `isActive` (and auxiliary methods) should not go to a different class. That would make it clear that `HaplotypeCallerEngine` is for calling regions and a different thing is for deciding active/non active